# Missing Value Ratio
## Formula:
### Ratio of missing value = no. of missing values / total no. of observations

In [9]:

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('dark_background')
import warnings
warnings.filterwarnings(action = 'ignore')

In [10]:
data = pd.read_csv("E:\Machine_Learning_Journey\Machine_Learning_Journey\data_sets\Transformed_Housing_Data2.csv")
data.head()

Sale_Price  No of Bedrooms  No of Bathrooms  Flat Area (in Sqft)  \
0    221900.0               3             1.00               1180.0   
1    538000.0               3             2.25               2570.0   
2    180000.0               2             1.00                770.0   
3    604000.0               4             3.00               1960.0   
4    510000.0               3             2.00               1680.0   

   Lot Area (in Sqft)  No of Floors  No of Times Visited  Overall Grade  \
0              5650.0           1.0                    0              7   
1              7242.0           2.0                    0              7   
2             10000.0           1.0                    0              6   
3              5000.0           1.0                    0              7   
4              8080.0           1.0                    0              8   

   Area of the House from Basement (in Sqft)  Basement Area (in Sqft)  ...  \
0                                     1180.0                        0  ...   
1                                     2170.0                      400  ...   
2                                      770.0                        0  ...   
3                                     1050.0                      910  ...   
4                                     1680.0                        0  ...   

   Waterfront_View_Yes  Zipcode_Group_Zipcode_Group_1  \
0                    0                              0   
1                    0                              0   
2                    0                              0   
3                    0                              0   
4                    0                              0   

   Zipcode_Group_Zipcode_Group_2  Zipcode_Group_Zipcode_Group_3  \
0                              0                              0   
1                              1                              0   
2                              1                              0   
3                              0                              1   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_4  Zipcode_Group_Zipcode_Group_5  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              1                              0   

   Zipcode_Group_Zipcode_Group_6  Zipcode_Group_Zipcode_Group_7  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_8  Zipcode_Group_Zipcode_Group_9  
0                              0                              0  
1                              0                              0  
2                              0                              0  
3                              0                              0  
4                              0                              0  

[5 rows x 31 columns]

In [11]:
data['No of Times Visited'] = np.where(data['No of Times Visited']==0, 
                                       np.nan ,data['No of Times Visited'] )

In [13]:
data['No of Times Visited'].head()

0   NaN
1   NaN
2   NaN
3   NaN
4   NaN
Name: No of Times Visited, dtype: float64

In [14]:
data['No of Times Visited'].isnull().sum()/len(data)

0.9017076218242399

In [17]:
data['No of Times Visited'].isnull()

0        True
1        True
2        True
3        True
4        True
         ... 
21604    True
21605    True
21606    True
21607    True
21608    True
Name: No of Times Visited, Length: 21609, dtype: bool

In [18]:
data['Years Since Renovation'].value_counts()

Years Since Renovation
0     20695
3        83
11       34
4        32
14       32
      ...  
58        1
70        1
55        1
65        1
69        1
Name: count, Length: 74, dtype: int64

### In above output 0 occurs 20695 times, 3 occurs 83 times and so on... // value_counts() function used in above

In [19]:
data['Years Since Renovation'] = np.where(data['Years Since Renovation']==0 ,
                                           np.nan, data['Years Since Renovation'])
data['Years Since Renovation'].isnull().sum()

20695

In [20]:
data.head()

Sale_Price  No of Bedrooms  No of Bathrooms  Flat Area (in Sqft)  \
0    221900.0               3             1.00               1180.0   
1    538000.0               3             2.25               2570.0   
2    180000.0               2             1.00                770.0   
3    604000.0               4             3.00               1960.0   
4    510000.0               3             2.00               1680.0   

   Lot Area (in Sqft)  No of Floors  No of Times Visited  Overall Grade  \
0              5650.0           1.0                  NaN              7   
1              7242.0           2.0                  NaN              7   
2             10000.0           1.0                  NaN              6   
3              5000.0           1.0                  NaN              7   
4              8080.0           1.0                  NaN              8   

   Area of the House from Basement (in Sqft)  Basement Area (in Sqft)  ...  \
0                                     1180.0                        0  ...   
1                                     2170.0                      400  ...   
2                                      770.0                        0  ...   
3                                     1050.0                      910  ...   
4                                     1680.0                        0  ...   

   Waterfront_View_Yes  Zipcode_Group_Zipcode_Group_1  \
0                    0                              0   
1                    0                              0   
2                    0                              0   
3                    0                              0   
4                    0                              0   

   Zipcode_Group_Zipcode_Group_2  Zipcode_Group_Zipcode_Group_3  \
0                              0                              0   
1                              1                              0   
2                              1                              0   
3                              0                              1   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_4  Zipcode_Group_Zipcode_Group_5  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              1                              0   

   Zipcode_Group_Zipcode_Group_6  Zipcode_Group_Zipcode_Group_7  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_8  Zipcode_Group_Zipcode_Group_9  
0                              0                              0  
1                              0                              0  
2                              0                              0  
3                              0                              0  
4                              0                              0  

[5 rows x 31 columns]

In [21]:
data['Basement Area (in Sqft)'] = np.where( data['Basement Area (in Sqft)'] ==0 ,
                                           np.nan, data['Basement Area (in Sqft)'])
data['Basement Area (in Sqft)'].isnull().sum()

13123

In [22]:
data.head()

Sale_Price  No of Bedrooms  No of Bathrooms  Flat Area (in Sqft)  \
0    221900.0               3             1.00               1180.0   
1    538000.0               3             2.25               2570.0   
2    180000.0               2             1.00                770.0   
3    604000.0               4             3.00               1960.0   
4    510000.0               3             2.00               1680.0   

   Lot Area (in Sqft)  No of Floors  No of Times Visited  Overall Grade  \
0              5650.0           1.0                  NaN              7   
1              7242.0           2.0                  NaN              7   
2             10000.0           1.0                  NaN              6   
3              5000.0           1.0                  NaN              7   
4              8080.0           1.0                  NaN              8   

   Area of the House from Basement (in Sqft)  Basement Area (in Sqft)  ...  \
0                                     1180.0                      NaN  ...   
1                                     2170.0                    400.0  ...   
2                                      770.0                      NaN  ...   
3                                     1050.0                    910.0  ...   
4                                     1680.0                      NaN  ...   

   Waterfront_View_Yes  Zipcode_Group_Zipcode_Group_1  \
0                    0                              0   
1                    0                              0   
2                    0                              0   
3                    0                              0   
4                    0                              0   

   Zipcode_Group_Zipcode_Group_2  Zipcode_Group_Zipcode_Group_3  \
0                              0                              0   
1                              1                              0   
2                              1                              0   
3                              0                              1   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_4  Zipcode_Group_Zipcode_Group_5  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              1                              0   

   Zipcode_Group_Zipcode_Group_6  Zipcode_Group_Zipcode_Group_7  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_8  Zipcode_Group_Zipcode_Group_9  
0                              0                              0  
1                              0                              0  
2                              0                              0  
3                              0                              0  
4                              0                              0  

[5 rows x 31 columns]

In [23]:
data.to_csv("NaN_Housing_Prices.csv", index = False)

# Actual work start here:

In [24]:
data = pd.read_csv('NaN_Housing_Prices.csv')
data.head()

Sale_Price  No of Bedrooms  No of Bathrooms  Flat Area (in Sqft)  \
0    221900.0               3             1.00               1180.0   
1    538000.0               3             2.25               2570.0   
2    180000.0               2             1.00                770.0   
3    604000.0               4             3.00               1960.0   
4    510000.0               3             2.00               1680.0   

   Lot Area (in Sqft)  No of Floors  No of Times Visited  Overall Grade  \
0              5650.0           1.0                  NaN              7   
1              7242.0           2.0                  NaN              7   
2             10000.0           1.0                  NaN              6   
3              5000.0           1.0                  NaN              7   
4              8080.0           1.0                  NaN              8   

   Area of the House from Basement (in Sqft)  Basement Area (in Sqft)  ...  \
0                                     1180.0                      NaN  ...   
1                                     2170.0                    400.0  ...   
2                                      770.0                      NaN  ...   
3                                     1050.0                    910.0  ...   
4                                     1680.0                      NaN  ...   

   Waterfront_View_Yes  Zipcode_Group_Zipcode_Group_1  \
0                    0                              0   
1                    0                              0   
2                    0                              0   
3                    0                              0   
4                    0                              0   

   Zipcode_Group_Zipcode_Group_2  Zipcode_Group_Zipcode_Group_3  \
0                              0                              0   
1                              1                              0   
2                              1                              0   
3                              0                              1   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_4  Zipcode_Group_Zipcode_Group_5  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              1                              0   

   Zipcode_Group_Zipcode_Group_6  Zipcode_Group_Zipcode_Group_7  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_8  Zipcode_Group_Zipcode_Group_9  
0                              0                              0  
1                              0                              0  
2                              0                              0  
3                              0                              0  
4                              0                              0  

[5 rows x 31 columns]

In [27]:
  data.head().shape

(5, 31)

In [28]:
data.isnull().sum()

Sale_Price                                       0
No of Bedrooms                                   0
No of Bathrooms                                  0
Flat Area (in Sqft)                              0
Lot Area (in Sqft)                               0
No of Floors                                     0
No of Times Visited                          19485
Overall Grade                                    0
Area of the House from Basement (in Sqft)        0
Basement Area (in Sqft)                      13123
Age of House (in Years)                          0
Latitude                                         0
Longitude                                        0
Living Area after Renovation (in Sqft)           0
Lot Area after Renovation (in Sqft)              0
Years Since Renovation                       20695
Condition_of_the_House_Excellent                 0
Condition_of_the_House_Fair                      0
Condition_of_the_House_Good                      0
Condition_of_the_House_Okay    

In [29]:
for i in data.columns:
  tmp = data[i].isnull().sum()/len(data)
  if tmp > 0.70:
    print("Removing column '"+i+"' with missing value ratio", tmp)
    data = data.drop(columns = [i])

Removing column 'No of Times Visited' with missing value ratio 0.9017076218242399
Removing column 'Years Since Renovation' with missing value ratio 0.9577028090147623


In [30]:
data.head()

Sale_Price  No of Bedrooms  No of Bathrooms  Flat Area (in Sqft)  \
0    221900.0               3             1.00               1180.0   
1    538000.0               3             2.25               2570.0   
2    180000.0               2             1.00                770.0   
3    604000.0               4             3.00               1960.0   
4    510000.0               3             2.00               1680.0   

   Lot Area (in Sqft)  No of Floors  Overall Grade  \
0              5650.0           1.0              7   
1              7242.0           2.0              7   
2             10000.0           1.0              6   
3              5000.0           1.0              7   
4              8080.0           1.0              8   

   Area of the House from Basement (in Sqft)  Basement Area (in Sqft)  \
0                                     1180.0                      NaN   
1                                     2170.0                    400.0   
2                                      770.0                      NaN   
3                                     1050.0                    910.0   
4                                     1680.0                      NaN   

   Age of House (in Years)  ...  Waterfront_View_Yes  \
0                       63  ...                    0   
1                       67  ...                    0   
2                       85  ...                    0   
3                       53  ...                    0   
4                       31  ...                    0   

   Zipcode_Group_Zipcode_Group_1  Zipcode_Group_Zipcode_Group_2  \
0                              0                              0   
1                              0                              1   
2                              0                              1   
3                              0                              0   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_3  Zipcode_Group_Zipcode_Group_4  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              1                              0   
4                              0                              1   

   Zipcode_Group_Zipcode_Group_5  Zipcode_Group_Zipcode_Group_6  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_7  Zipcode_Group_Zipcode_Group_8  \
0                              0                              0   
1                              0                              0   
2                              0                              0   
3                              0                              0   
4                              0                              0   

   Zipcode_Group_Zipcode_Group_9  
0                              0  
1                              0  
2                              0  
3                              0  
4                              0  

[5 rows x 29 columns]

# Low Variance Code Technique 2:

In [36]:
data = pd.read_csv("E:\Machine_Learning_Journey\Machine_Learning_Journey\data_sets\Raw_Housing_Prices.csv")
data.head()

ID Date House was Sold  Sale Price  No of Bedrooms  \
0  7129300520     14 October 2017    221900.0               3   
1  6414100192    14 December 2017    538000.0               3   
2  5631500400    15 February 2016    180000.0               2   
3  2487200875    14 December 2017    604000.0               4   
4  1954400510    15 February 2016    510000.0               3   

   No of Bathrooms  Flat Area (in Sqft)  Lot Area (in Sqft)  No of Floors  \
0             1.00               1180.0              5650.0           1.0   
1             2.25               2570.0              7242.0           2.0   
2             1.00                770.0             10000.0           1.0   
3             3.00               1960.0              5000.0           1.0   
4             2.00               1680.0              8080.0           1.0   

  Waterfront View No of Times Visited  ... Overall Grade  \
0              No                 NaN  ...             7   
1              No                 NaN  ...             7   
2              No                 NaN  ...             6   
3              No                 NaN  ...             7   
4              No                 NaN  ...             8   

   Area of the House from Basement (in Sqft)  Basement Area (in Sqft)  \
0                                     1180.0                        0   
1                                     2170.0                      400   
2                                      770.0                        0   
3                                     1050.0                      910   
4                                     1680.0                        0   

   Age of House (in Years)  Renovated Year  Zipcode  Latitude  Longitude  \
0                       63               0  98178.0   47.5112   -122.257   
1                       67            1991  98125.0   47.7210   -122.319   
2                       85               0  98028.0   47.7379   -122.233   
3                       53               0  98136.0   47.5208   -122.393   
4                       31               0  98074.0   47.6168   -122.045   

   Living Area after Renovation (in Sqft)  Lot Area after Renovation (in Sqft)  
0                                  1340.0                                 5650  
1                                  1690.0                                 7639  
2                                  2720.0                                 8062  
3                                  1360.0                                 5000  
4                                  1800.0                                 7503  

[5 rows x 21 columns]

In [37]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   ID                                         21613 non-null  int64  
 1   Date House was Sold                        21613 non-null  object 
 2   Sale Price                                 21609 non-null  float64
 3   No of Bedrooms                             21613 non-null  int64  
 4   No of Bathrooms                            21609 non-null  float64
 5   Flat Area (in Sqft)                        21604 non-null  float64
 6   Lot Area (in Sqft)                         21604 non-null  float64
 7   No of Floors                               21613 non-null  float64
 8   Waterfront View                            21613 non-null  object 
 9   No of Times Visited                        2124 non-null   object 
 10  Condition of the House

In [38]:
numerical_data = data.drop(columns = ['ID', 'Date House was Sold', 'Sale Price',
                               'Waterfront View', 'No of Times Visited',
                            'Condition of the House', 'Overall Grade', 'Zipcode'
                              ])

numerical_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 13 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   No of Bedrooms                             21613 non-null  int64  
 1   No of Bathrooms                            21609 non-null  float64
 2   Flat Area (in Sqft)                        21604 non-null  float64
 3   Lot Area (in Sqft)                         21604 non-null  float64
 4   No of Floors                               21613 non-null  float64
 5   Area of the House from Basement (in Sqft)  21610 non-null  float64
 6   Basement Area (in Sqft)                    21613 non-null  int64  
 7   Age of House (in Years)                    21613 non-null  int64  
 8   Renovated Year                             21613 non-null  int64  
 9   Latitude                                   21612 non-null  float64
 10  Longitude             

In [43]:
from sklearn.preprocessing import Normalizer
from sklearn.impute import SimpleImputer
# Step 1: Handle missing values using SimpleImputer
imputer = SimpleImputer(strategy='mean')  # You can use 'median', 'most_frequent', or 'constant'
numerical_data_imputed = pd.DataFrame(imputer.fit_transform(numerical_data), columns=numerical_data.columns)

# Step 2: Apply normalization
norm = Normalizer()
norm_data = norm.fit_transform(numerical_data_imputed)

# Step 3: Convert the normalized data back to a DataFrame
norm_data = pd.DataFrame(norm_data, columns=numerical_data.columns)

# Display the first few rows of the normalized data
norm_data.head()

No of Bedrooms  No of Bathrooms  Flat Area (in Sqft)  Lot Area (in Sqft)  \
0        0.000363         0.000121             0.142628            0.682924   
1        0.000264         0.000198             0.226174            0.637335   
2        0.000152         0.000076             0.058440            0.758961   
3        0.000527         0.000395             0.258159            0.658569   
4        0.000263         0.000175             0.147072            0.707346   

   No of Floors  Area of the House from Basement (in Sqft)  \
0      0.000121                                   0.142628   
1      0.000176                                   0.190972   
2      0.000076                                   0.058440   
3      0.000132                                   0.138299   
4      0.000088                                   0.147072   

   Basement Area (in Sqft)  Age of House (in Years)  Renovated Year  Latitude  \
0                 0.000000                 0.007615        0.000000  0.005743   
1                 0.035202                 0.005896        0.175219  0.004200   
2                 0.000000                 0.006451        0.000000  0.003623   
3                 0.119859                 0.006981        0.000000  0.006259   
4                 0.000000                 0.002714        0.000000  0.004169   

   Longitude  Living Area after Renovation (in Sqft)  \
0  -0.014777                                0.161968   
1  -0.010765                                0.148729   
2  -0.009277                                0.206437   
3  -0.016121                                0.179131   
4  -0.010684                                0.157577   

   Lot Area after Renovation (in Sqft)  
0                             0.682924  
1                             0.672273  
2                             0.611874  
3                             0.658569  
4                             0.656834

In [44]:
tmp = norm_data.var()
tmp

No of Bedrooms                               4.072961e-08
No of Bathrooms                              2.802522e-08
Flat Area (in Sqft)                          1.231845e-02
Lot Area (in Sqft)                           1.352561e-02
No of Floors                                 3.012314e-08
Area of the House from Basement (in Sqft)    1.010425e-02
Basement Area (in Sqft)                      1.647478e-03
Age of House (in Years)                      2.001937e-05
Renovated Year                               1.683169e-03
Latitude                                     1.078556e-05
Longitude                                    7.113676e-05
Living Area after Renovation (in Sqft)       1.079669e-02
Lot Area after Renovation (in Sqft)          1.251729e-02
dtype: float64

In [45]:
T = 1e-04

for i in tmp:
  if i < T:
    norm_data.drop(columns = [tmp[tmp ==i].index[0]], inplace = True)

norm_data.var()

Flat Area (in Sqft)                          0.012318
Lot Area (in Sqft)                           0.013526
Area of the House from Basement (in Sqft)    0.010104
Basement Area (in Sqft)                      0.001647
Renovated Year                               0.001683
Living Area after Renovation (in Sqft)       0.010797
Lot Area after Renovation (in Sqft)          0.012517
dtype: float64